Perftests scripts (#427)
mezo-org/mezod: docs/performance-tests.md

and re-checking of all pending transactions can consume substantial node resources, potentially
affecting overall network performance during sustained high-demand periods

For users, this distinction means that transaction submission strategies must be more
carefully thought on Mezo than on Ethereum. Transactions submitted with marginally
competitive gas prices that might eventually clear on Ethereum could be completely removed
from consideration on Mezo, requiring resubmission with higher fees.

#### Gas Limit per transaction and block sizing

Mezo takes a different approach to determining which transactions fit into a block compared
to Ethereum, particularly regarding block size calculations and gas consumption mechanics.

In Mezo, the effective space in a block is determined by a fixed gas limit per block, similar
to Ethereum. However, Mezo's transaction selection process differs significantly in how
it accounts for gas consumption:

##### Gas Accounting Differences

While Ethereum charges based on the actual computational resources used during execution (with
an upfront maximum specified by the user's gasLimit), Mezo implements a minimum gas consumption

> rule. Specifically, Mezo will consume at least half of the gasLimit specified in a transaction,

Reviewer: This parameter is configurable. 50% is just the default value. Worth mentioning it.

Author: fair point. I'll add a line about it

Author: addressed in 90cbdc8ea7b82c777fbc258fd2d03d5c96631b75

  its transaction pruning behavior, as excessive mempool growth triggers more
  frequent re-checking operations that can impact node performance. Persistent
  mempool growth may indicate the need for increased block capacity or fee market
  adjustments.

- Transactions per blocks: An easier way to understand the gas consumed per block.

- Time take per block: How long validators takes to construct a valid block,
  including transaction selection, execution, and state updates. Extended block
  building times can lead to inconsistent block production intervals and
  reduced overall throughput (Mezo is trying to achieve block production every
  3.5 seconds in average). This measurement is especially relevant when
  evaluating how the re-checking behavior of the mempool affects validator
  performance under load.

- Gas price: This represent the cost per unit of gas, and reflect the current
  congestion level of the network. Under load the gas price will increase in an
  exponential manner, and decrease when the congestion of the network reduce.

### Scripts

The scripts used to run these tests are available in the mezod repository

> [here](https://github.com/mezo-org/mezod/tree/main/tests/perfs).

Reviewer: Broken link ser

Author: correct fixed in: a59aa802860012975aa8494de3f3db12abc426a1, (it's still gonna be broken 'til merge though)

  including transaction selection, execution, and state updates. Extended block
  building times can lead to inconsistent block production intervals and
  reduced overall throughput (Mezo is trying to achieve block production every
  3.5 seconds in average). This measurement is especially relevant when
  evaluating how the re-checking behavior of the mempool affects validator
  performance under load.

- Gas price: This represent the cost per unit of gas, and reflect the current
  congestion level of the network. Under load the gas price will increase in an
  exponential manner, and decrease when the congestion of the network reduce.

### Scripts

The scripts used to run these tests are available in the mezod repository
[here](https://github.com/mezo-org/mezod/tree/main/tests/performance).

### Tests results

All tests are run for a period of 3 minutes, and bots will send transactions
in constant flow to the node.

Note that for all these transaction, the gasLimit per transaction was set to:

- native transfers: 21000 gas
- ERC20 precompile transfers: 25000
- ERC20 transfers: 25000

For both the native and ERC20 precompile transfers the whole gas would be consumed.

> Fore the ERC20 transfes, between 21000 and 24000 of gas was used, the rest would be

Reviewer: ```suggestion
Fore the ERC20 transfers, between 21000 and 24000 of gas was used, the rest would be
```

refunded.

This behaviour is the reason driving the transaction count per blocks accross the
different transactions types.

#### Performances comparison of the transactions

> The following charts shows execution at maximum capacity of the block, without

Reviewer: ```suggestion
The following charts show execution at maximum capacity of the block, without
```

too much congestion on the node. The node always produce full blocks
and the backlog is small enough / large enough to produce a full new next block.
We can easily see the natural increase of the gas price. As soon as the flow of transaction
reduce, the gas price is affected and lowered.

> The number of transaction in each blocks are similar with a maximum of 476 for the
> native transfer, and 400 for the others.

Reviewer: ```suggestion
The numbers of transactions in each block are similar: maximum of 476 for the
native transfer, and 400 for the others.
```

> These charts shows the blocks of 10m gas, for the 3 transfers types under similar loads

Reviewer: ```suggestion
These charts show the blocks of 10m gas, for the 3 transfers types under similar loads
```

as described previously.

##### Native transfers

![10m txs](./assets/10m_native_txs.png)
![10m gas](./assets/10m_native_gas.png)

##### ERC20 precompile transfers

![10m txs](./assets/10m_precompile_txs.png)
![10m gas](./assets/10m_precompile_gas.png)

##### ERC20 transfers

![10m txs](./assets/10m_erc20_txs.png)
![10m gas](./assets/10m_erc20_gas.png)

Note: the next steps results can be considered for any of the transactions.

#### 10m gas per blocks

#### 15m gas per blocks

#### 17.5m gas per blocks

> At 17.5m gas per block, the network can process around 700 transactions per blocks.

Reviewer: ```suggestion
At 17.5m gas per block, the network can process around 700 transactions per block.
```

##### Under low stress

In the following, the node processed 44k transation in 180 seconds (an average of 244
transaction per seconds).

The time take to produce block stayed stable, and the mempool
size slowly increased as the node was able execute transaction sufficiently quickly.

The gas price incerease nicely other time to support the volume of transaction.
The node was running close to 100% CPU, but we do not really observe a slow down.

![175m gas block txs](./assets/175m_low_txs.png)
![175m gas block gas](./assets/175m_low_gas.png)

##### Under higher stress

Here sent a lot more order straight away, filling up the mempool rather quickly. The first

> effect observe is the time taken to execute every blocks (as high as 8 seconds per blocks),

Reviewer: ```suggestion
effect observed is the time taken to execute every blocks (as high as 8 seconds per blocks),
```
